File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/graphql/Cookies.java
package com.zappymart.adapters.graphql;

import org.springframework.http.ResponseCookie;

import java.time.Duration;

public final class Cookies {

    public static final String GRAPHQL_PATH = "/graphql";

    public static final Duration LIFETIME = Duration.ofDays(30);

    private Cookies() {
    }

    public static ResponseCookie cart(String cartId, boolean secure) {
        return ResponseCookie.from(SecuritySettings.CART_COOKIE, cartId)
                .httpOnly(true)
                .secure(secure)
                .sameSite("Lax")
                .path("/")
                .maxAge(LIFETIME)
                .build();
    }

    public static ResponseCookie refresh(String refreshToken, boolean secure) {
        return ResponseCookie.from(SecuritySettings.REFRESH_COOKIE, refreshToken)
                .httpOnly(true)
                .secure(secure)
                .sameSite("Lax")
                .path(GRAPHQL_PATH)
                .maxAge(LIFETIME)
                .build();
    }

    public static ResponseCookie clearedRefresh(boolean secure) {
        return ResponseCookie.from(SecuritySettings.REFRESH_COOKIE, "")
                .httpOnly(true)
                .secure(secure)
                .sameSite("Lax")
                .path(GRAPHQL_PATH)
                .maxAge(Duration.ZERO)
                .build();
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/graphql/Cookies.kt
package nl.zappymart.adapters.graphql

import java.time.Duration
import nl.zappymart.domain.accounts.Session
import org.springframework.http.ResponseCookie

class Cookies(private val properties: SecurityProperties) {

    fun refreshToken(value: String): String = ResponseCookie.from(REFRESH_TOKEN_NAME, value)
        .httpOnly(true)
        .secure(properties.cookiesSecure)
        .sameSite("Lax")
        .path(properties.graphQlPath)
        .maxAge(Duration.ofDays(Session.LIFETIME_IN_DAYS))
        .build()
        .toString()

    fun clearedRefreshToken(): String = ResponseCookie.from(REFRESH_TOKEN_NAME, "")
        .httpOnly(true)
        .secure(properties.cookiesSecure)
        .sameSite("Lax")
        .path(properties.graphQlPath)
        .maxAge(Duration.ZERO)
        .build()
        .toString()

    fun cart(cartId: String): String = ResponseCookie.from(CART_NAME, cartId)
        .httpOnly(true)
        .secure(properties.cookiesSecure)
        .sameSite("Lax")
        .path("/")
        .maxAge(Duration.ofDays(CART_LIFETIME_IN_DAYS))
        .build()
        .toString()

    companion object {
        const val REFRESH_TOKEN_NAME = "zappy_refresh"
        const val CART_NAME = "zappy_cart"
        const val CART_LIFETIME_IN_DAYS = 30L
    }
}
